File: frontend/src/api/identify.ts
import { apiClient } from './client'
import type {
  IdentifyCandidate,
  IdentifyFace,
  IdentifyFaceResult,
  IdentifyRequestQuery,
  IdentifyResponse,
} from './types'

interface BackendCandidate {
  rank: number
  faceId: string | null
  personId: string | null
  sampleId: string | null
  name: string | null
  score: number
  decision: string
  cropImageUrl: string | null
}

interface BackendFace {
  queryFaceId: string
  boundingBox: { x: number; y: number; width: number; height: number }
  qualityScore: number | null
  result: BackendCandidate | null
  candidates: BackendCandidate[]
}

interface BackendIdentifyResponse {
  requestId: string
  status: string
  decision: string | null
  faceCount: number | null
  queryImageUrl: string | null
  faces: BackendFace[]
  createdAt: string
  completedAt: string | null
}

function normalizeCandidate(c: BackendCandidate): IdentifyCandidate {
  return {
    rank: c.rank,
    faceId: c.faceId ?? '',
    personId: c.personId ?? '',
    sampleId: c.sampleId ?? '',
    name: c.name || undefined,
    score: c.score,
    decision: c.decision as IdentifyCandidate['decision'],
    cropImageUrl: c.cropImageUrl || undefined,
  }
}

export function normalizeIdentifyResponse(raw: unknown): IdentifyResponse {
  const r = raw as BackendIdentifyResponse

  const faces: IdentifyFace[] = (r.faces ?? []).map((f) => {
    const best = f.result ?? f.candidates[0]
    const result: IdentifyFaceResult = {
      status: (best?.decision || r.decision || 'no_match') as IdentifyFaceResult['status'],
      personId: best?.personId ?? '',
      faceId: best?.faceId ?? '',
      sampleId: best?.sampleId ?? '',
      name: best?.name || undefined,
      score: best?.score ?? 0,
      threshold: 0,
    }
    return {
      queryFaceId: f.queryFaceId,
      boundingBox: f.boundingBox,
      qualityScore: f.qualityScore ?? 0,
      result,
      candidates: f.candidates.map(normalizeCandidate),
    }
  })

  return {
    requestId: r.requestId,
    status: r.status,
    decision: r.decision || '',
    faceCount: r.faceCount ?? 0,
    queryImageUrl: r.queryImageUrl || '',
    faces,
    createdAt: r.createdAt,
  }
}

export function identify(file: File, options?: IdentifyRequestQuery): Promise<IdentifyResponse> {
  const formData = new FormData()
  formData.append('image', file)

  const searchParams = new URLSearchParams()
  if (options?.topK !== undefined) searchParams.set('topK', String(options.topK))
  if (options?.selectedFaceIndex !== undefined) searchParams.set('selectedFaceIndex', String(options.selectedFaceIndex))
  if (options?.threshold !== undefined) searchParams.set('threshold', String(options.threshold))

  const query = searchParams.toString()
  const url = query ? `/identify?${query}` : '/identify'

  return apiClient
    .post<BackendIdentifyResponse>(url, formData, {
      headers: {
        'Content-Type': 'multipart/form-data',
      },
    })
    .then((res) => normalizeIdentifyResponse(res.data))
}

File: backend/app/schemas/identify.py
"""Identification request/response schemas."""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.common import ListingResponse


class BoundingBox(BaseModel):
    x: int
    y: int
    width: int
    height: int


class Candidate(BaseModel):
    rank: int
    faceId: UUID | None
    personId: UUID | None
    sampleId: UUID | None
    name: str | None
    score: float
    decision: str
    cropImageUrl: str | None = None


class IdentifyFaceResult(BaseModel):
    queryFaceId: UUID
    boundingBox: BoundingBox
    qualityScore: float | None
    result: Candidate | None
    candidates: list[Candidate]


class IdentifyResponse(BaseModel):
    requestId: UUID
    status: str
    decision: str | None
    faceCount: int | None
    queryImageUrl: str | None
    faces: list[IdentifyFaceResult]
    createdAt: datetime
    completedAt: datetime | None


class IdentificationRequestSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    requestId: UUID
    status: str
    decision: str | None
    faceCount: int | None
    topK: int
    createdAt: datetime
    completedAt: datetime | None


class IdentificationRequestListResponse(ListingResponse[IdentificationRequestSummary]):
    pass


class IdentifyQueryParams(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    topK: int = Field(default=5, ge=1, le=20)
    selectedFaceIndex: int | None = Field(default=None, ge=0)
    threshold: float | None = Field(default=None, ge=0.0, le=1.0)
